velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -33,129 +33,125 @@ std::string getResultScale(std::string precision, std::string scale) {
+  VELOX_DCHECK(!overflow);
+  return r;
 }
 
-// Both x_value and y_value must be >= 0.
+// Add two non-negative values by adding the whole and fraction parts
+// separately.
 template <typename R, typename A, typename B>
-inline static R addLargePositive(
-    const A& a,
-    const B& b,
+inline static R addLargeNonNegative(
+    const A a,
+    const B b,
     uint8_t aScale,
     uint8_t bScale,
     uint8_t rScale) {
-  VELOX_DCHECK_GE(a, 0);
-  VELOX_DCHECK_GE(b, 0);
+  VELOX_USER_CHECK_GE(
+      a, 0, "Non-negative value is expected in addLargeNonNegative.");
+  VELOX_USER_CHECK_GE(
+      b, 0, "Non-negative value is expected in addLargeNonNegative.");
 
-  // Separate out whole/fractions.
-  auto [aLeft, aRight] = getWholeAndFraction<A>(a, aScale);
-  auto [bLeft, bRight] = getWholeAndFraction<B>(b, bScale);
+  // Separate whole and fraction parts.
+  const auto [aWhole, aFraction] = getWholeAndFraction<A>(a, aScale);
+  const auto [bWhole, bFraction] = getWholeAndFraction<B>(b, bScale);
 
   // Adjust fractional parts to higher scale.
   auto higherScale = std::max(aScale, bScale);
-  int128_t aRightScaled =
-      checkAndIncreaseScale<A>(aRight, higherScale - aScale);
-  int128_t bRightScaled =
-      checkAndIncreaseScale<B>(bRight, higherScale - bScale);
-
-  R right;
-  int64_t carryToLeft;
-  auto multiplier = velox::DecimalUtil::kPowersOfTen[higherScale];
-  if (aRightScaled >= multiplier - bRightScaled) {
-    right = R(aRightScaled - (multiplier - bRightScaled));
+  int128_t aFractionScaled = increaseScale<A>(aFraction, higherScale - aScale);
+  int128_t bFractionScaled = increaseScale<B>(bFraction, higherScale - bScale);
+
+  R fraction;
+  int64_t carryToLeft = 0;
+  auto carrier = velox::DecimalUtil::kPowersOfTen[higherScale];
+  if (aFractionScaled >= carrier - bFractionScaled) {
+    fraction = R(aFractionScaled + bFractionScaled - carrier);
     carryToLeft = 1;
   } else {
-    right = R(aRightScaled + bRightScaled);
-    carryToLeft = 0;
+    fraction = R(aFractionScaled + bFractionScaled);
   }
-  right = checkAndReduceScale<R>(R(right), higherScale - rScale);
+  fraction = reduceScale<R>(R(fraction), higherScale - rScale);
 
-  auto left = R(aLeft) + R(bLeft) + R(carryToLeft);
-  return R(left * velox::DecimalUtil::kPowersOfTen[rScale]) + R(right);
+  auto whole = R(aWhole) + R(bWhole) + R(carryToLeft);
+  return R(whole * velox::DecimalUtil::kPowersOfTen[rScale]) + R(fraction);
 }
 
-/// A and b cannot be 0, and one must be positive and the other
-/// negative.
+// Add two opposite values.
 template <typename R, typename A, typename B>
-inline static R addLargeNegative(
-    const A& a,
-    const B& b,
+inline static R addLargeOpposite(
+    const A a,
+    const B b,
     uint8_t aScale,
     uint8_t bScale,
     int32_t rScale) {
-  VELOX_DCHECK_NE(a, 0);
-  VELOX_DCHECK_NE(b, 0);
-  VELOX_DCHECK((a < 0 && b > 0) || (a > 0 && b < 0));
+  VELOX_USER_CHECK(
```

Comment:
This logic is guaranteed by this `else` branch, so it is `VELOX_DCHECK_XX`.

https://github.com/facebookincubator/velox/pull/5791/commits/0578c4cbe3f6260ec0f726e2fef7ba7b442397cf#diff-4baee9f82f9c347f753972415d345941d2c2a110b608d480b090650171da096bL156

Reply:
Fixed.